### ISSUE TITLE
Wrong configuration info is used for a C/C++ file that is included by another

### ISSUE THREAD

**Issue by Volker-Weissmann** [microsoft/vscode-cpptools#6250]
I was/I am trying to make "Go To Definition" work in a big project but it did not work. I managed to produce a minimal example of this bug:

**Type: LanguageService**

**Describe the bug**
- OS and Version: Arch Linux
- VS Code Version: 
1.49.1
58bb7b2331731bf72587010e943852e13e6fd3cf
x64
- C/C++ Extension Version: 1.0.0
- Other extensions you installed (and if the issue persists after disabling them): None
- Does this issue involve using SSH remote to run the extension on a remote machine?: No
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc):

File Structure (You can download it here: [bugexample.zip](https://github.com/microsoft/vscode-cpptools/files/5322388/bugexample.zip)):
.
├── child.C  : #include "child.H"
├── child.H  : #include "hiddenFile.H"
├── compile_commands.json
├── hiddenlib
│   └── hiddenFile.H : empty
├── main.C : #include "child.H" \n  int main(){}
└── .vscode
│   └── c_cpp_properties.json

c_cpp_properties.json:
```
{
    "configurations": [
      {
        "compileCommands": "${workspaceFolder}/compile_commands.json",
      }
    ],
    "version": 4
}
```
compile_commands.json was generated using bear -- g++ -I./hiddenlib -c main.C:
```
[
  {
    "arguments": [
      "/usr/bin/g++",
      "-I./hiddenlib",
      "-c",
      "main.C"
    ],
    "directory": "/home/volker/Documents/bugexample",
    "file": "/home/volker/Documents/bugexample/main.C",
    "output": "/home/volker/Documents/bugexample/main.o"
  }
]
```

**Expected behavior and observed behaviour**

If I open the folder, open child.H, set my cursor on "hiddenFile.H" and sellect "Go To Definition" it should jump to hiddenlib/hiddenFile.H. Instead it says "No definition for hiddenFile". Also there is a message on startup that says 
`"child.C" not found in "${workspaceFolder}/compile_commands.json". 'includePath' from c_cpp_properties.json in folder 'bugexample' will be used for this file instead.`

Note: You can only reproduce this bug if you don't open main.C and you need to wait a few seconds after startup before trying to "Go To Definition" (around 7 seconds on my machine).

<details>
  <summary><strong>Logs</strong></summary>

```
-------- Diagnostics - 10/4/2020, 12:46:46 AM
Version: 1.0.0
Current Configuration:
{
    "compileCommands": "${workspaceFolder}/compile_commands.json",
    "compilerArgs": [],
    "browse": {
        "limitSymbolsToIncludedHeaders": true
    }
}
Translation Unit Mappings:
[ /home/volker/Documents/bugexample/child.C ]:
    /home/volker/Documents/bugexample/child.H
Translation Unit Configurations:
[ /home/volker/Documents/bugexample/child.C ]:
    Process ID: 66481
    Memory Usage: 9 MB
    Compiler Path: /usr/bin/clang
    Includes:
        /usr/include/c++/10.2.0
        /usr/include/c++/10.2.0/x86_64-pc-linux-gnu
        /usr/include/c++/10.2.0/backward
        /usr/local/include
        /usr/lib/clang/10.0.1/include
        /usr/include
        /home/volker/Documents/bugexample
    Standard Version: c++14
    IntelliSense Mode: clang-x64
    Other Flags:
        --clang
        --clang_version=100001
Total Memory Usage: 9 MB

------- Workspace parsing diagnostics -------
Number of folders and files enumerated: 64661
Number of files discovered (not excluded): 60879

```
</details>

Slightly OT: If you know a better tool than vscode-cpptools for finding the definitions in big projects, please tell me.

**Comment by sean-mcmanus** [comment]
This is "by design". child.H is being associated with the child.C TU. After you open child.H, you can run C/C++: Log Diagnostics to see what TU is being used. You should be able to fix the issue via adding the the hiddenlib folder to the includePath that is used for that TU, i.e. via adding an entry for child.C to compile_commands.json or adding hiddenlib to an includePath property in c_cpp_properties.json. Is that sufficient, or did you have another solution in mind?

**Comment by Volker-Weissmann** [comment]
The compile_commands.json file says that child.H and child.C are not the same TU. Shouldn't the compile_commands.json file be the ultimate authority? It should be if we want "Go To Definition" to find the code that is actually executed. I assume that 

> adding the the hiddenlib folder to the includePath that is used for that TU, i.e. via adding an entry for child.C to compile_commands.json or adding hiddenlib to an includePath property in c_cpp_properties.json

would work, but would require manual labor on every file, which is a bad thing. [Clangd](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd) does it right, it takes compile_commands.json as the ultimate authority and therefore works in this case. I can't see anything that is better in your approach than in clangd. If you think that its bad coding style that child.H and child.C are not in the same TU, you should add a note to the description of your plugin (or at the documentation of the "compileCommands": "...json" setting) that this only works if the code follows certain coding standards.


Also, the fact that "Go to definition" works as soon as I open main.C feels like a bug.

**Comment by sean-mcmanus** [comment]
The compile_commands.json you provided doesn't mention child.C:
```
[
  {
    "arguments": [
      "/usr/bin/g++",
      "-I./hiddenlib",
      "-c",
      "main.C"
    ],
    "directory": "/home/volker/Documents/bugexample",
    "file": "/home/volker/Documents/bugexample/main.C",
    "output": "/home/volker/Documents/bugexample/main.o"
  }
]
```
Are you referring to a different file?

Go to definition with main.C works, because it's included in your compile_commands.json to use -I./hiddenlib.

**Comment by Volker-Weissmann** [comment]
In the real-life example of this bug, child.H is something like this
```
#include "hiddenFile.H"
#ifndef CHILD_H
#define CHILD_H
#include "child.C"
#endif
```

You see that compile_commands.json doesn't mention child.C, because g++ child.C is never executed.

**Comment by sean-mcmanus** [comment]
Yeah, we don't currently handle the case where .C files are included.

**Comment by Volker-Weissmann** [comment]
> Yeah, we don't currently handle the case where .C files are included.

Then you should document [here](https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference) that it only works if no .C files are included. Because the C++ standard allows including .C files.
